feat: document the conjunction/disjunction convention (#885)
leanprover-community/leanprover-community.github.io: templates/contribute/style.md

in the source as well.  The lambda notation `λ x ↦ x * x`, while syntactically valid,
is disallowed in mathlib in favor of the `fun` keyword.

### Conjunctions, disjunctions

Hypotheses should typically not be conjoined nor disjoined because it makes the lemma harder to use:
```lean
-- Instead of...
example (hPQ : P ∧ Q) : R := ...

example (hST : S ∨ T) : U := ...

example (hV : ∃ i, V i) : W := ...

-- ...do
example (hP : P) (hQ : Q) : R := ...

example (hS : S) : U := ...
example (hT : T) : U := ...

example {i} (hV : V i) : W := ...

> ```

Reviewer: ````suggestion
Hypotheses should not be conjunctions, as this typically makes the lemma harder to use:
```lean
-- Instead of
example (hPQ : P ∧ Q) : R := ...
-- do
example (hP : P) (hQ : Q) : R := ...
```

In the majority of situations, hypotheses should not be disjunctions or existentials, for the same reason:
```lean
-- Instead of
example (hST : S ∨ T) : U := ...

example (hV : ∃ i, V i) : W := ...
-- do
example (hS : S) : U := ...
example (hT : T) : U := ...

example {i} (hV : V i) : W := ...
```
````

Reviewer: My intention is to make it clearer that the conjunction rule is a rule, and the disjunction/existential is a guideline.

Similarly, the return type of a lemma should not be a conjunction.
One should instead prove two lemmas.

> It is acceptable to prove the two lemmas from a conjunction lemma if that reduces code duplication.

Reviewer: ```suggestion
Similarly, the return type of a lemma should not be a conjunction, and one should instead prove two lemmas.
It is acceptable to prove the two lemmas from a private conjunction lemma if that reduces code duplication.
```

Finally, an existential result can be turned using choice into a definition along with a lemma

> about that definition. Whether this is a sensible change to make highly depends on the situation.

Reviewer 2: ```suggestion
Finally, an existential result can be turned, using choice, into a definition along with a lemma about that definition. Whether this is a sensible change to make depends highly on the situation.
```

some missing commas and wrong word order.

Author: This might be a German-English difference?

Reviewer: I agree that the commas should be added. To me, "depends highly" and "highly depends" are both clear though, but "depends highly" feels more natural.

Author: ```suggestion
Finally, an existential result can be turned, using choice, into a definition along with a lemma
about that definition. Whether this is a sensible change to make depends highly on the situation.
```

> about that definition. Whether this is a sensible change to make depends highly on the situation.

Reviewer: ```suggestion
Finally, an existential result can be turned, using choice, into a definition along with a lemma about that definition. 
Whether this is a sensible change to make depends highly on the situation.
```
optional, but this follows the line-breaking in the rest of your PR!

in the source as well.  The lambda notation `λ x ↦ x * x`, while syntactically valid,
is disallowed in mathlib in favor of the `fun` keyword.

### Conjunctions, disjunctions

Hypotheses should not be conjunctions, as this typically makes the lemma harder to use:
```lean
-- Instead of
example (hPQ : P ∧ Q) : R := ...
-- do
example (hP : P) (hQ : Q) : R := ...
```

Similarly, the return type of a lemma should not be a conjunction, and one should instead prove two lemmas.

> It is acceptable to prove the two lemmas from a private conjunction lemma if that reduces code duplication.

Reviewer: ```suggestion
It is acceptable, however, to prove the two lemmas from a private conjunction lemma if that reduces code duplication.
```
Minor but this makes it clear that this is an allowed exception, rather than an explanation or clarification of the previous line.

```lean
-- Instead of
example (hPQ : P) : Q ∧ R := ...
-- do
example (hP : P) : Q := ...
example (hP : P) : R := ...
```

In the majority of situations, hypotheses should not be disjunctions, for the same reason:
```lean
-- Instead of
example (hST : S ∨ T) : U := ...
-- do
example (hS : S) : U := ...
example (hT : T) : U := ...
```

> Exceptions can be made when abiding to this rule would result in many lemmas with similar function:

Reviewer: ```suggestion
Exceptions can be made when abiding to this rule would result in many very similar lemmas:
```
optional, but I think I like the phrasing in your example more

```lean
-- This is acceptable because the alternative would be to write four very similar lemmas
lemma ENNReal.inv_div {a b : ENNReal} (htop : b ≠ ⊤ ∨ a ≠ ⊤) (hzero : b ≠ 0 ∨ a ≠ 0) :
    (a / b)⁻¹ = b / a := ...
```

A similar transformation could be made to existential hypotheses:
```lean
-- Instead of
example (hV : ∃ i, V i) : W := ...
-- one could do
example {i} (hV : V i) : W := ...
```

> This is mostly discouraged as providing `i` explicitly can be harder than proving its existence.

Reviewer: I think "mostly discouraged" might be a bit strong, eg there's a bunch of lemmas in mathlib which take `{i} (hV : V i)` and I don't want someone modifying them all to take existentials. Can we use phrasing similar to the conclusion one below - where we mention the option but don't enforce one way or the other?

Reviewer: (In particular, this is now the negation of what it was earlier, which is what Michael called uncontroversial!)
